src/sagemaker/jumpstart/curated_hub/curated_hub.py

```diff
@@ -113,3 +182,236 @@ def delete_model(self, model_name: str, model_version: str = "*") -> None:
     def delete(self) -> None:
         """Deletes this Curated Hub"""
         return self._sagemaker_session.delete_hub(self.hub_name)
+
+    def _is_invalid_model_list_input(self, model_list: List[Dict[str, str]]) -> bool:
+        """Determines if input args to ``sync`` is correct.
+
+        `model_list` objects must have `model_id` (str) and optional `version` (str).
+        """
+        for obj in model_list:
+            if not isinstance(obj.get("model_id"), str):
+                return True
+            if "version" in obj and not isinstance(obj["version"], str):
+                return True
+        return False
+
+    def _populate_latest_model_version(self, model: Dict[str, str]) -> Dict[str, str]:
+        """Populates the lastest version of a model from specs no matter what is passed.
+
+        Returns model ({ model_id: str, version: str })
+        """
+        model_specs = utils.verify_model_region_and_return_specs(
+            model["model_id"], "*", JumpStartScriptScope.INFERENCE, self.region
+        )
+        return {"model_id": model["model_id"], "version": model_specs.version}
+
+    def _get_jumpstart_models_in_hub(self) -> List[Dict[str, Any]]:
+        """Returns list of `HubContent` that have been created from a JumpStart model."""
+        hub_models = self.list_models()
+
+        js_models_in_hub = []
+        for hub_model in hub_models["HubContentSummaries"]:
+            # TODO: extract both in one pass
+            jumpstart_model_id = next(
+                (
+                    tag
+                    for tag in hub_model["search_keywords"]
+                    if tag.startswith(JUMPSTART_HUB_MODEL_ID_TAG_PREFIX)
+                ),
+                None,
+            )
+            jumpstart_model_version = next(
+                (
+                    tag
+                    for tag in hub_model["search_keywords"]
+                    if tag.startswith(JUMPSTART_HUB_MODEL_VERSION_TAG_PREFIX)
+                ),
+                None,
+            )
+
+            if jumpstart_model_id and jumpstart_model_version:
+                js_models_in_hub.append(hub_model)
+
+        return js_models_in_hub
+
+    def _determine_models_to_sync(
+        self, model_list: List[JumpStartModelInfo], models_in_hub: Dict[str, Any]
+    ) -> List[JumpStartModelInfo]:
+        """Determines which models from `sync` params to sync into the CuratedHub.
+
+        Algorithm:
+
+            First, look for a match of model name in Hub. If no model is found, sync that model.
+
+            Next, compare versions of model to sync and what's in the Hub. If version already
+            in Hub, don't sync. If newer version in Hub, don't sync. If older version in Hub,
+            sync that model.
+        """
+        models_to_sync = []
+        for model in model_list:
```

Comment:
nit: Consider using python set `.filter()` to remove models that shouldn't be synced